### ISSUE TITLE
RADICAL has a number of issues--needs overhaul

### ISSUE THREAD

**Issue by cstork9** [mlpack/mlpack#1004]
radical.cpp  seems to have severe problems:  
1) Produces nan result when m==0, which is the default
2) Algorithm appears to have an algorithm bug in Vasicek() since log( neg number ) is a common occurance
3)  20x-100x slower than it needs to be
4) The timers make it not thread safe
5) Doesn't work on complex valued matrices

The test code just checks for consistent results, not correct results.  

This is a cool algorithm, if it works.....
Seems like this code hasn't been tested properly and isn't production capable.

**Comment by rcurtin** [comment]
Hi there,

This is not really a good bug report because it completely lacks details.

1. I am not seeing this behavior nor can I reproduce it so it must be dependent on your dataset or the parameters you are using.  If you can provide more information, maybe I can help.
2. The only case in which this is possible would be when all (or some) elements of `z` have the exact same value.  I would be careful about referring to this as a bug, because the paper simply makes no consideration of this possibility.  I can't reproduce this situation either, I think because of the noise that is added to the data in `DoRadical2D()`.
3. According to whom?  I can see a few possible opportunities for acceleration, but there is no baseline for how fast it could be.  Probably it is possible to avoid copying each data dimension for the sort, and probably also possible to sort once, but I am not sure.  The sort is expensive and difficult to work around though.
4. I could agree that the internal timers in the method should be removed, but meh, if your timers overwrite themselves it's not going to break the results.
5. This one would be very tricky to implement.  Why not simply take the real and complex dimensions as two dimensions in a non-complex matrix?

Testing RADICAL or any machine learning algorithm fully is a very difficult prospect.  If you have a nice idea for how to test it "properly" and guarantee that it works---feel free to propose it.  But any machine learning algorithm that has any randomness is particularly difficult to test for correctness.  We can only test that it is not "extremely wrong".

**Comment by cstork9** [comment]
Thanks.

I would say this code has more problems than just bugs.  It's just a poor
implementation that needs to be improved.  I may do it.

Christof

**Comment by rcurtin** [comment]
Hi Christof,

I did not write the implementation though I did review it in depth some years back; I believe it to be a correct implementation of RADICAL as per the paper.  However, I am still not able to reproduce the problem you are having, so I think it is a stretch to call it a 'poor implementation' based only on the problems you're reporting here.  Note that nobody else who has used this code has reported any issue (granted, I don't have any statistics on who is using what).  So I think that your problems probably lie more with the specific dataset you are using, which exposes an edge case that the code doesn't currently handle properly.

If you can provide information on that, then I can at least set aside some time (when I have a chance) to figure out what your issue is there.  Otherwise there is nothing I can do.

That said, improvements are of course welcome---no code is perfect.  If you have good ideas for refactoring the code, improving the runtime, or writing new tests, I am happy to help out.

**Comment by cstork9** [comment]
OK, thanks.  Sounds reasonable.

I am surprised that you haven't gotten more comments on the code.  That's
good feedback.  I'll check on my implementation.

I may make some proposed improvement.  Testing this code is hard, partially
because a good test takes hours to run.

Christof

**Comment by rcurtin** [comment]
Yes, so I noticed a few things when I glanced at it again.  The data matrix is expanded with the `replicates`  option.  So if you keep that number low, it runs definitely faster.  In addition, the sorting procedure in `Vasicek()` is necessarily going to be slow since it will incur so much memory reordering.  Maybe each dimension, after replicates have been added, can be sorted only once at the beginning of the algorithm.  Also I suspect that perhaps we can avoid copying the dataset `replicates` times, and possibly only add the noise when it is needed.  But that would require some thinking to ensure that it doesn't break the algorithm itself.
